File: function_examples/read_write_example.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdarg.h>
#include "opendal.h"

/*

- Read: Full read from a path with no configuration

- Write: Full write to a path with no configuration (resets when writing to the same path)

*/

opendal_bytes create_bytes(char* str) {
    return (opendal_bytes) {
        .data = (uint8_t*)str,
        .len = strlen(str)
    };
}

// OpenDAL is not guaranteed to null-terminate messages
void safe_printf_opendal_messages(const char* fmt, ...) {
    va_list args;
    va_start(args, fmt);

    while (*fmt) {
        if (*fmt == '%') { // Only handle %s and %B (opendal_bytes)
            fmt++;
            if (*fmt == 's') {
                const char* str = va_arg(args, const char*);
                fputs(str, stdout);
            } else {
                assert(*fmt == 'B');
                opendal_bytes bytes = va_arg(args, opendal_bytes);
                printf("%.*s", (int)bytes.len, bytes.data);
            }
            fmt++;
        } else {
            putchar(*fmt++);
        }
    }

    va_end(args);
}

// Test1: Read before writing (read from non-existent path)
void test1(opendal_operator *op, char *path) {
    opendal_result_read r = opendal_operator_read(op, path);
    assert(r.error != NULL);
    
    printf("Error: %u\n", r.error->code);
    opendal_bytes error_msg = r.error->message;
    safe_printf_opendal_messages("Error message: %B\n", error_msg);
    
    opendal_error_free(r.error);
}

// Test2: Write and read
void test2(opendal_operator *op, char *path, opendal_bytes data) {
    // Write
    opendal_error* error = opendal_operator_write(op, path, &data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path (%s): %B\n", path, data);

    // Read
    opendal_result_read r = opendal_operator_read(op, path);
    assert(r.error == NULL);
    opendal_bytes read_data = r.data;
    assert(read_data.len == data.len);
    safe_printf_opendal_messages("Read from path (%s): %B\n", path, read_data);

    // Cleanup (Necessary to free the error when it exists)
    opendal_bytes_free(&read_data);
}

// Test3: Double write (bigger before smaller data)
void test3(opendal_operator *op, char *path, opendal_bytes data, opendal_bytes bigger_data) {
    opendal_error* error = opendal_operator_write(op, path, &bigger_data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path (%s): %B\n", path, bigger_data);

    error = opendal_operator_write(op, path, &data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path (%s): %B\n", path, data);

    opendal_result_read r = opendal_operator_read(op, path);
    assert(r.error == NULL);
    opendal_bytes read_data = r.data;
    safe_printf_opendal_messages("Read from path (%s): %B\n", path, read_data);

    opendal_bytes_free(&read_data);
}

// Test4: Write to multiple paths (chained) and special characters
void test4(opendal_operator *op, char *path1, char *path2, char *path3, opendal_bytes data, opendal_bytes bigger_data, opendal_bytes special_data) {
    // Write to paths
    opendal_error* error = opendal_operator_write(op, path1, &data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path1 (%s): %B\n", path1, data);
    error = opendal_operator_write(op, path2, &bigger_data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path2 (%s): %B\n", path2, bigger_data);
    error = opendal_operator_write(op, path3, &special_data);
    assert(error == NULL);
    safe_printf_opendal_messages("Wrote to path3 (%s): %B\n", path3, special_data);

    // Read from paths
    opendal_result_read r1 = opendal_operator_read(op, path1);
    opendal_result_read r2 = opendal_operator_read(op, path2);
    opendal_result_read r3 = opendal_operator_read(op, path3);
    assert(r1.error == NULL && r2.error == NULL && r3.error == NULL);

    // Check data
    opendal_bytes read_data1 = r1.data;
    opendal_bytes read_data2 = r2.data;
    opendal_bytes read_data3 = r3.data;
    assert(read_data1.len == data.len && read_data2.len == bigger_data.len && read_data3.len == special_data.len);
    safe_printf_opendal_messages("Read from path1 (%s): %B\n", path1, read_data1);
    safe_printf_opendal_messages("Read from path2 (%s): %B\n", path2, read_data2);
    safe_printf_opendal_messages("Read from path3 (%s): %B\n", path3, read_data3);

    // Cleanup
    opendal_bytes_free(&read_data1);
    opendal_bytes_free(&read_data2);
    opendal_bytes_free(&read_data3);
}

// Test5: Write bigger data (2^20 = 1048576 bytes = 1 MB)
void test5(opendal_operator *op, char *path) {
    int fd = open("function_examples/bytes.txt", O_RDONLY);
    assert(fd != -1);
    int size = 1048576;
    char buffer[size];

    int res = read(fd, &buffer[0], size);
    assert(res == size);
    close(fd);
    
    opendal_bytes data = {
        .data = (uint8_t*)buffer,
        .len = size
    };

    opendal_error* error = opendal_operator_write(op, path, &data);
    assert(error == NULL);
    printf("First 8 bytes written: %.8s\n", data.data);
    printf("Last 8 bytes written: %.8s\n", &data.data[size - 8]);

    opendal_result_read r = opendal_operator_read(op, path);
    assert(r.error == NULL);
    opendal_bytes read_data = r.data;
    assert(read_data.len == data.len);
    printf("First 8 bytes read: %.8s\n", read_data.data);
    printf("Last 8 bytes read: %.8s\n", &read_data.data[size - 8]);
    printf("Read data has the expected length: %zu\n", read_data.len);

    opendal_bytes_free(&read_data);
}

int main(void) {
    // Data
    char *str = "Test message";
    char *bigger_str = "A bigger text with more characters";
    char *special_str = "Text with some special characters: !@#$%^&*()_+☺";

    // Paths
    char *path = "/testpath";
    char *path2 = "/testpath/secondpath";
    char *path3 = "/testpath/testpath";

    // Create operator with memory backend
    opendal_result_operator_new result = opendal_operator_new("memory", NULL);
    assert(result.op != NULL);
    assert(result.error == NULL);

    // Get operator and data
    opendal_operator *op = result.op;
    opendal_bytes data = create_bytes(str);
    opendal_bytes bigger_data = create_bytes(bigger_str);
    opendal_bytes special_data = create_bytes(special_str);

    // Tests
    printf("\n------------ Test1: read non-existent path ---------------------\n\n");
    test1(op, path);
    printf("\n------------ Test2: write and read -----------------------------\n\n");
    test2(op, path, data);
    printf("\n------------ Test3: double write -------------------------------\n\n");
    test3(op, path, data, bigger_data);
    printf("\n------------ Test4: chained paths and special chars ------------\n\n");
    test4(op, path, path2, path3, data, bigger_data, special_data);
    printf("\n------------ Test5: write 1MB of data --------------------------\n\n");
    test5(op, path);
    printf("\n----------------------------------------------------------------\n\n");

    // Cleanup
    opendal_operator_free(op);
    return 0;
}

File: Rust/examples/read_write_example.rs
use opendal::Result;
use opendal::layers::LoggingLayer;
use opendal::services;
use opendal::Operator;
use opendal::Buffer;

// Path Constants
const TEST_PATH: &str = "/testpath";
const TEST_PATH_SECOND: &str = "/testpath/secondpath";
const TEST_PATH_NESTED: &str = "/testpath/testpath";

fn create_bytes(data: &str) -> Vec<u8> {
    data.as_bytes().to_vec()
}

fn format_bytes(data: &[u8]) -> String {
    String::from_utf8_lossy(data).to_string()
}

fn format_buffer(buffer: &Buffer) -> String {
    let bytes: Vec<u8> = buffer.to_bytes().to_vec();
    String::from_utf8_lossy(&bytes).to_string()
}

async fn write_with_logging(op: &Operator, path: &str, data: &[u8]) -> Result<()> {
    let data_owned = data.to_vec();
    match op.write(path, data_owned).await {
        Ok(_) => {
            println!("Wrote to path ({}): {}", path, format_bytes(data));
            Ok(())
        }
        Err(e) => {
            println!("Failed to write to path ({}): {:?}", path, e);
            Err(e)
        }
    }
}

async fn read_with_logging(op: &Operator, path: &str) -> Result<Buffer> {
    match op.read(path).await {
        Ok(data) => {
            println!("Read from path ({}): {}", path, format_buffer(&data));
            Ok(data)
        }
        Err(e) => {
            println!("Failed to read from path ({}): {:?}", path, e);
            Err(e)
        }
    }
}

///////////////////////////////////////////////////////////////////////////////////////////

async fn test1(op: &Operator, path: &str) -> Result<()> {
    let result = op.read(path).await;
    match result {
        Ok(_) => {
            println!("Expected an error, but got data");
            panic!("Expected an error when reading non-existent path");
        }
        Err(e) => {
            println!("Received expected error: {:?}", e);
            Ok(())
        }
    }
}

async fn test2(op: &Operator, path: &str, data: &[u8]) -> Result<()> {
    write_with_logging(op, path, data).await?;
    read_with_logging(op, path).await?;
    Ok(())
}

async fn test3(op: &Operator, path: &str, small_data: &[u8], big_data: &[u8]) -> Result<()> {
    // Write bigger data
    write_with_logging(op, path, big_data).await?;
    // Write smaller data
    write_with_logging(op, path, small_data).await?;
    // Read the result
    read_with_logging(op, path).await?;
    Ok(())
}

async fn test4(op: &Operator, path1: &str, path2: &str, path3: &str, data: &[u8], bigger_data: &[u8], special_data: &[u8]) -> Result<()> {
    // Write to paths
    write_with_logging(op, path1, data).await?;
    write_with_logging(op, path2, bigger_data).await?;
    write_with_logging(op, path3, special_data).await?;
    // Read from paths
    read_with_logging(op, path1).await?;
    read_with_logging(op, path2).await?;
    read_with_logging(op, path3).await?;
    Ok(())
}

///////////////////////////////////////////////////////////////////////////////////////////

#[tokio::main]
async fn main() -> Result<()> {
    let builder = services::Memory::default();

    let op = Operator::new(builder)?
        .layer(LoggingLayer::default())
        .finish();

    let data = create_bytes("Test message");
    let bigger_data = create_bytes("A bigger text with more characters");
    let special_data = create_bytes("Text with some special characters: !@#$%^&*()_+☺");

    println!("\n------------ Test1: read non-existent path ---------------------\n");
    test1(&op, TEST_PATH).await?;
    println!("\n------------ Test2: write and read -----------------------------\n");
    test2(&op, TEST_PATH, &data).await?;
    println!("\n------------ Test3: double write -------------------------------\n");
    test3(&op, TEST_PATH, &data, &bigger_data).await?;
    println!("\n------------ Test4: chained paths and special chars ------------\n");
    test4(&op, TEST_PATH,TEST_PATH_SECOND, TEST_PATH_NESTED, &data, &bigger_data, &special_data).await?;
    println!("\n----------------------------------------------------------------\n");
    Ok(())
}
